stan/math/rev/functor/coupled_ode_system.hpp

```diff
@@ -438,53 +142,68 @@ struct coupled_ode_system<F, var, var> {
    *    expected number of derivatives, N.
    */
   void operator()(const std::vector<double>& z, std::vector<double>& dz_dt,
-                  double t) const {
+                  double t) {
     using std::vector;
 
+    dz_dt.resize(size());
+
     // Run nested autodiff in this scope
     nested_rev_autodiff nested;
 
-    vector<var> y_vars(z.begin(), z.begin() + N_);
+    /*auto local_args_tuple_ = apply([&](auto&&... args) {
+        return
+       std::tuple<decltype(deep_copy_vars(args))...>(deep_copy_vars(args)...);
+        }, args_tuple_);*/
 
-    vector<var> dy_dt_vars = f_(t, y_vars, theta_nochain_, x_, x_int_, msgs_);
+    Eigen::Matrix<var, Eigen::Dynamic, 1> y_vars(N_);
+    for (size_t n = 0; n < N_; ++n)
+      y_vars.coeffRef(n) = z[n];
 
-    check_size_match("coupled_ode_system", "dz_dt", dy_dt_vars.size(), "states",
+    Eigen::Matrix<var, Eigen::Dynamic, 1> f_y_t_vars
+        = apply([&](auto&&... args) { return f_(t, y_vars, msgs_, args...); },
+                local_args_tuple_);
+
+    check_size_match("coupled_ode_system", "dy_dt", f_y_t_vars.size(), "states",
                      N_);
 
-    for (size_t i = 0; i < N_; i++) {
-      dz_dt[i] = dy_dt_vars[i].val();
-      dy_dt_vars[i].grad();
+    for (size_t i = 0; i < N_; ++i) {
+      dz_dt[i] = f_y_t_vars.coeffRef(i).val();
+      f_y_t_vars.coeffRef(i).grad();
+
+      y_adjoints_ = y_vars.adj();
+
+      memset(args_adjoints_.data(), 0, sizeof(double) * args_vars_);
+      apply(
+          [&](auto&&... args) {
+            accumulate_adjoints(args_adjoints_.data(), args...);
+          },
+          local_args_tuple_);
+
+      apply([&](auto&&... args) { zero_local_adjoints(args...); },
+            local_args_tuple_);
 
-      for (size_t j = 0; j < N_; j++) {
+      nested.set_zero_all_adjoints();
```

Comment:
I would avoid zeroing on the last pass for I==n-1 when its not needed. Avoiding cycling over the tape is a good thing.